### PR TITLE
Mechanism for adding/removing "tags" to/from tensor names, after changes to the tensor

### LINKED ISSUE #454: Implement a mechanism for adding/removing "tags" to/from tensor names, after changes to the tensor
🐞 **What bugs you?**

Every Tensor Object has a `name` parameter, that stores the user-defined (and modifiable) name. After any change is made to the tensor, for example a Lorentz Transform or a Config Change, the name remains unmodified. This may make tracking changes to the tensor difficult for extended calculations.

🎯 **Goal**

A tag added to the Tensor name, after **binary** changes are made to the Tensor to mark the respective changes, would be nice to have. For example, after change_config or lorentz_transformation, tags, like "_cc" or "_lt" can be appended/removed.

💡 **Possible solutions**

Write a helper function for altering `name` conditionally.

Edit `BaseRelativityTensor.lorentz_transform()`.

Edit `change_config()` for all Tensors in `einsteinpy.symbolic`.

📋  **Steps to solve the problem**

 * Comment below about what you've started working on.
 * Add, commit, push your changes
 * Submit a pull request and add this in comments - `Addresses #<put issue number here>`
 * Ask for a review in comments section of pull request
 * Celebrate your contribution to this project 🎉

### PR DESCRIPTION
**Pull request by JeS24**
Fixes #454 

@ritzvik I have made a few slightly unrelated changes, which my code depends on. Please review the PR. Thanks.

### CODE REVIEW

**Comment by JeS24** [comment]
I can elaborate on the "unrelated changes", if you want me to, but those are really few in number. Also, do you want me put the new helper functions in `einstein.symbolic.tensor` or leave it in `einstein.symbolic.helpers.`?

**Comment by ritzvik** [comment]
> I can elaborate on the "unrelated changes", if you want me to, but those are really few in number. Also, do you want me put the new helper functions in `einstein.symbolic.tensor` or leave it in `einstein.symbolic.helpers.`?

I'll see to the changes myself and ask if explanation is required.

`helpers` it is.

**Comment by ritzvik** [comment]
I'll see this tomorrow, it's netflix and quarantine time :D

**Comment by JeS24** [comment]
> I'll see this tomorrow, it's netflix and quarantine time :D

Ah, alright. I want to get rid of the comments in the helper function. So, when you're done and if there's no major issue, please tell me. I will make another commit.

**Comment by ritzvik** [comment]
Also, If you are aiming for GSoC, it's high time to start with a proposal :)

**Comment by JeS24** [comment]
> Also, If you are aiming for GSoC, it's high time to start with a proposal :)

COVID derailed my initial plans. So, I have rebooted the entire proposal. Will share soon.

File: src/einsteinpy/symbolic/einstein.py [hunk 1]
```diff
@@ -10,7 +11,9 @@ class EinsteinTensor(BaseRelativityTensor):
     Class for defining Einstein Tensor
     """
 
-    def __init__(self, arr, syms, config="ll", parent_metric=None):
+    def __init__(
+        self, arr, syms, config="ll", parent_metric=None, name="EinsteinTensor"
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/einstein.py]:
```suggestion
        self, arr, syms, config="ll", parent_metric=None, name=None
```
Let this remain `None` for now. Why discrimination between `ChristoffelSymbols` and `EinsteinTensor` ? 
😃

Comment (ritzvik) [reply src/einsteinpy/symbolic/einstein.py]: User can inspect the class to get the Tensor sub-type. No need for default `EinsteinTensor` in my opinion

File: src/einsteinpy/symbolic/einstein.py [hunk 2]
```diff
@@ -25,6 +28,8 @@ def __init__(self, arr, syms, config="ll", parent_metric=None):
         parent_metric : ~einsteinpy.symbolic.metric.MetricTensor or None
             Corresponding Metric for the Einstein Tensor.
             Defaults to None.
+        name : String
+            Name of the Tensor. Defaults to "EinsteinTensor". To be modified, only when the Tensor object is modified in some way
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/einstein.py]:
```suggestion
            Name of the Tensor. Defaults to `None`.
```

File: src/einsteinpy/symbolic/helpers.py [hunk 3]
```diff
@@ -187,3 +188,106 @@ def inv(self):
...
+    Function to get all possible combinations of lower (l) and upper (u) indices for a general tensor
+    
+    Parameters
+    ----------
+    rank : int or None
+           Rank of the Tensor, used with config_change. Defaults to None.
+    
+    Returns
+    -------
+    ~list : Contains all possible combinations of lower and upper indices for a given rank
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/helpers.py]:
```suggestion
    list
        Contains all possible combinations of lower and upper indices for a given rank
```

File: src/einsteinpy/symbolic/helpers.py [hunk 4]
```diff
@@ -187,3 +188,106 @@ def inv(self):
         tmp_array = np.reciprocal(tmp_array)
         self._inv = ImmutableDenseNDimArray(tmp_array)
         return self._inv
+
+
+def get_config_combinations(rank: int) -> list:
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/helpers.py]:
```suggestion
def _get_config_combinations(rank: int) -> list:
```
This is not a user-facing function. Let's make this hidden

**Comment by ritzvik** [comment]
Firstly, what I think, the changes for a not-so-complicated issue are overly complicated. I got a hang of what you are tying to do, but it would be good if you could give an overview.

From what I guess, you are tying to track changes, `lt` or `cc`, `cc` being tracked by the index of the specific configuration in a list of all possible configurations of a given rank. But I think it would be unnecessary, as the user can readily see the config by `tensor.config`. Also, methods of tracking the number of changes is questionable as `__` is something the user can put in the name of the tensor since it was initiated, albeit highly improbable. 

I would propose very simple solution. If the name is `None`, no need for change. If the tensor is changed in a order `lt->ull->lt->luu`. Just make the name to be `TensorName->lt->ull->lt->luu`. No need to replace previous `lt`s and `cc`s and make the code complicated. And please drop the `config_combinations` thing. It is neither intuitive to the user, nor to future code contributors.

File: src/einsteinpy/symbolic/einstein.py (hunk 1 above)
Comment (ritzvik) [reply src/einsteinpy/symbolic/einstein.py]: Okay, I would take back my suggestion. Let it remain as it is

**Comment by ritzvik** [comment]
Another suggestion, No tensor name defaults to `None`

For `BaseRelativityTensor`, the base class, let `name="GenericTensor"` be the default.

**Comment by JeS24** [comment]
> From what I guess, you are tying to track changes, `lt` or `cc`, `cc` being tracked by the index of the specific configuration in a list of all possible configurations of a given rank. But I think it would be unnecessary, as the user can readily see the config by `tensor.config`. Also, methods of tracking the number of changes is questionable as `__` is something the user can put in the name of the tensor since it was initiated, albeit highly improbable.  I would propose very simple solution. If the name is `None`, no need for change. If the tensor is changed in a order `lt->ull->lt->luu`. Just make the name to be `TensorName->lt->ull->lt->luu`. No need to replace previous `lt`s and `cc`s and make the code complicated. 

This is what, I intended to do in the beginning. The problem, I encountered was that most users (I believe) will be using this in REPL mode in Jupyter, where, running the same code block (containing `change_config()` and/or `lorentz_transform()` ) again and again to test or rerun portions of code, can cause the name to grow very quickly and can add redundant tags to the name, defeating the purpose of adding easy to read tags. But I am aware, that the current implementation is not optimal. I will work on refactoring the code.

> And please drop the `config_combinations` thing. It is neither intuitive to the user, nor to future code contributors.

I agree. I separated this and the other helper function from `change_name`, just to get better scores on Cognitive and Cyclomatic Complexity, but as I said, I will work on a better implementation.

**Comment by JeS24** [comment]
@ritzvik How about tracking only `__lt` and appending `__newconfig` changes without any checks?`einsteinpy.symbolic.tensors > Tensors._change_config()` already prevents changing to the current config (i.e. if `self.config == newconfig`). We can track `__lt`, and remove or add it, as and when required, while we can simply append `__newconfig` to the `name`, each time. This implies giving up short names, but would add the ability to track all past config changes and overall, is an okay compromise between code complexity and intended behaviour, in my opinion. Although, the issue of names growing to be very long, still persists, but I do believe, most people would not mind it. This way, I guess, the function can be brought down to 3 `if`s.

**Comment by ritzvik** [comment]
@JeS24 sure, take your time

The jupyter argument you gave is reasonable, but an annoyed user is always free to reset the name of the tensor by `tensor.name = "blah blah"`

**Comment by ritzvik** [comment]
Also, do the needful when you finalize everything
https://github.com/einsteinpy/einsteinpy/wiki/How-to-refactor-commits

**Comment by JeS24** [comment]
> Another suggestion, No tensor name defaults to `None`

My thoughts exactly.

> For `BaseRelativityTensor`, the base class, let `name="GenericTensor"` be the default.

Will add this. Plus, can we extend it to other Tensor classes, like `MetricTensor`, with something like `GenericMetricTensor`? Also, `lorentz_transform()` currently resets the `parent_metric` to `None`, every time, it is called. This prevents further `config_change()` operations. Would it be fine to change `None` to `self.parent_metric`, given that, the base class, `BaseRelativityTensor` has this attribute.

**Comment by ritzvik** [comment]
Remove tracking altogether because Lorentz transform would be dropped as and when coordinate basis change gets implemented. No point in enhancing a feature which would be dropped in near future.

Regarding parent_metric gets forwarded as None, it is an pre-existing issue.

**Comment by JeS24** [comment]
> Remove tracking altogether because Lorentz transform would be dropped as and when coordinate basis change gets implemented. No point in enhancing a feature which would be dropped in near future.

I didn't know. Will do.

> Regarding parent_metric gets forwarded as None, it is an pre-existing issue.

Will discuss this soon.

**Comment by JeS24** [comment]
@ritzvik Meanwhile, there's an issue. `change_config()` in every tensor, except `MetricTensor`, has been implemented, such that the config is changed, irrespective of whether `newconfig == current_config`. As such, for MetricTensor, trying to change to the same config does not alter the `name`, while for all other tensors (including `Vector`, `ChristoffelSymbols`), it alters the `name` and appends the same config.

**Comment by ritzvik** [comment]
See #374 for relevant discussion

**Comment by ritzvik** [comment]
> @ritzvik Meanwhile, there's an issue. `change_config()` in every tensor, except `MetricTensor`, has been implemented, such that the config is changed, irrespective of whether `newconfig == current_config`. As such, for MetricTensor, trying to change to the same config does not alter the `name`, while for all other tensors (including `Vector`, `ChristoffelSymbols`), it alters the `name` and appends the same config.

Yes, partly because Metric Tensor can either assume “ll” or “uu” configuration, and both are inverse of each other. So it was handled differently.

**Comment by JeS24** [comment]
> > @ritzvik Meanwhile, there's an issue. `change_config()` in every tensor, except `MetricTensor`, has been implemented, such that the config is changed, irrespective of whether `newconfig == current_config`. As such, for MetricTensor, trying to change to the same config does not alter the `name`, while for all other tensors (including `Vector`, `ChristoffelSymbols`), it alters the `name` and appends the same config.
> 
> Yes, partly because Metric Tensor can either assume “ll” or “uu” configuration, and both are inverse of each other. So it was handled differently.

I get that, it's binary for metric tensors, but won't it be better to handle other tensors similarly? If the new config is exactly the same as the current config, why make a function call to `_change_config()` and return another object, or is there a check somewhere that prevents this and I cannot find it?

**Comment by JeS24** [comment]
> See #374 for relevant discussion

I see. I think, until General Coordinate Transforms are implemented, it's better to put the `parent_metric` issue on hold.

**Comment by JeS24** [comment]
> > > @ritzvik Meanwhile, there's an issue. `change_config()` in every tensor, except `MetricTensor`, has been implemented, such that the config is changed, irrespective of whether `newconfig == current_config`. As such, for MetricTensor, trying to change to the same config does not alter the `name`, while for all other tensors (including `Vector`, `ChristoffelSymbols`), it alters the `name` and appends the same config.
> > 
> > 
> > Yes, partly because Metric Tensor can either assume “ll” or “uu” configuration, and both are inverse of each other. So it was handled differently.
> 
> I get that, it's binary for metric tensors, but won't it be better to handle other tensors similarly? If the new config is exactly the same as the current config, why make a function call to `_change_config()` and return another object, or is there a check somewhere that prevents this and I cannot find it?

So, the only remaining issue is this. I have refactored `change_name()` and made edits at the places, you pointed out in the PR review.

**Comment by shreyasbapat** [comment]
There are issues with docstrings :)

File: src/einsteinpy/symbolic/helpers.py [hunk 5]
```diff
@@ -187,3 +188,106 @@ def inv(self):
...
+
+def change_name(current_name: str, context: str, rank=None) -> str:
+    """
+    Function to add descriptive tags to Tensor name parameter, after the tensor is modified.
+    Currently handles Lorentz Transformation and Config Changes.
+
+    Parameters
+    ----------
+    current_name : str
+                Current name of the tensor
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/helpers.py]: The tab spacing is not right

File: src/einsteinpy/symbolic/helpers.py [hunk 6]
```diff
@@ -187,3 +188,106 @@ def inv(self):
...
+                                   - current configuration for config_change (cc)
+                                   e.g. '__luu', '__ull' and other comnbinations for a Rank 3 Tensor
+    
+    rank : int or None
+           Rank of the Tensor, used with config_change only. Defaults to None.
+    
+    Returns
+    -------
+    altered_name : str
+                Altered name of the tensor, with appropriate tags
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/helpers.py]: Same here. Use 4 spaces.

File: src/einsteinpy/symbolic/helpers.py [hunk 7]
```diff
@@ -187,3 +188,106 @@ def inv(self):
...
+    current_config_index : int
+                           Index of current tensor configuration
+    
+    """
+    # Index of matching config in current_name among all configs
+    current_config_index = [conf in current_name for conf in configs_all].index(True)
+    return current_config_index
+
+
+def change_name(current_name: str, context: str, rank=None) -> str:
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/helpers.py]: Type specification is not needed here.

File: src/einsteinpy/symbolic/ricci.py [hunk 8]
```diff
@@ -28,6 +29,8 @@ def __init__(self, arr, syms, config="ll", parent_metric=None):
         parent_metric : ~einsteinpy.symbolic.metric.MetricTensor or None
             Corresponding Metric for the Ricci Tensor.
             Defaults to None.
+        name : String
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/ricci.py]:
```suggestion
        name : str
```

File: src/einsteinpy/symbolic/ricci.py [hunk 9]
```diff
@@ -28,6 +29,8 @@ def __init__(self, arr, syms, config="ll", parent_metric=None):
...
+            Name of the Tensor. Defaults to "RicciTensor". To be modified, only when the Tensor object is modified in some way
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/ricci.py]: You should add a TODO ? This is not the right way of doing this.

**Comment by shreyasbapat** [comment]
Also, I request you to write useful PR titles. So its not a nightmare for me while writing changelog.

**Comment by JeS24** [comment]
> There are issues with docstrings :)

Does `tox -e reformat` help with the proper formatting? Also, I have either modified or removed parts of the docstrings. So, most docstring issues should be sorted out now. I will commit the changes and ask for a PR review soon.

File: src/einsteinpy/symbolic/ricci.py (hunk 9 above)
Comment (JeS24) [reply src/einsteinpy/symbolic/ricci.py]: > Also, I request you to write useful PR titles. So its not a nightmare for me while writing changelog.

Sure. Will keep this in mind.

**Comment by JeS24** [comment]
> Also, I request you to write useful PR titles. So its not a nightmare for me while writing changelog.

Sure, will keep this in mind,

File: src/einsteinpy/symbolic/ricci.py (hunk 9 above)
Comment (JeS24) [reply src/einsteinpy/symbolic/ricci.py]: This was not a ToDo. I meant, the `name` will only be modified, when the tensor is modified in some way (Lorentz Transform or Config Change). I have removed this portion, regardless, as `name` is user-modifiable and we are not tracking changes to the tensor anymore.

**Comment by ritzvik** [comment]
> > There are issues with docstrings :)
> 
> Does `tox -e reformat` help with the proper formatting? Also, I have either modified or removed parts of the docstrings. So, most docstring issues should be sorted out now. I will commit the changes and ask for a PR review soon.

The command take care of the code formatting only.

> > Also, I request you to write useful PR titles. So its not a nightmare for me while writing changelog.
> 
> Sure, will keep this in mind,

You can always edit the title.

**Comment by ritzvik** [comment]
Whatever changes you deem to be fit, based on our discussion, please do it. We can have another round of discussions if needed :)

**Comment by shreyasbapat** [comment]
This is great. Please restrict the commits to 2/3. Rebase(if behind) and rename the PR.

**Comment by JeS24** [comment]
> This is great. Please restrict the commits to 2/3. Rebase(if behind) and rename the PR.

And I was typing to ask you, if merging 9 commits will be an issue, as using VSCode's Interactive Diff and Merge has made the entire working tree dirty and now I can't seem to merge or rebase. No worries. I will try and get it cleaned, after a while and update here. 😄

**Comment by shreyasbapat** [comment]
Use command line :p

**Comment by JeS24** [comment]
I am closing this PR and opening another one. There are some changes, that were made to tests and `metric.py`, that are causing cyclic conflicts. And, I don't know how to fix that.

**Pull request closed without merging**